### PR TITLE
LPD-39807

### PR DESCRIPTION
**Pull request by rijema**
https://liferay.atlassian.net/browse/LPD-39807

### CODE REVIEW

**Comment by rijema** [comment]
Previous: https://github.com/liferay-bpm/liferay-portal/pull/3769

**Comment by rijema** [comment]
ci:test:sf

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/registry/DDMServiceUpgradeStepRegistrator.java [hunk 1]
```diff
@@ -596,6 +596,17 @@ protected String[][] getTableAndPrimaryKeyColumnNames() {
...
 					_fileEntryFriendlyURLResolver, _groupLocalService,
 					_userLocalService));
+
+		registry.register(
+			"5.6.1", "5.7.0",
+			new BaseExternalReferenceCodeUpgradeProcess() {
+
+				@Override
+				protected String[][] getTableAndPrimaryKeyColumnNames() {
+					return new String[][] {{"DDMStructure", "structureId"}};
```
Comment (guilhermedcamacho) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/registry/DDMServiceUpgradeStepRegistrator.java:606]: Should we use `structureId` for `primKeyColumnName` here?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java [hunk 2]
```diff
@@ -92,6 +97,26 @@ public static void setUpClass() {
...
+	public void testAddDDMStructure() throws Exception {
+		DDMStructure ddmStructure = _addStructure(
+			RandomTestUtil.randomString());
+
+		Assert.assertTrue(
+			Validator.isNotNull(ddmStructure.getExternalReferenceCode()));
+	}
+
+	@Test(expected = DuplicateDDMStructureExternalReferenceCodeException.class)
+	public void testAddDDMStructureWithExistingExternalReferenceCode()
```
Comment (guilhermedcamacho) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java]: You should cover this scenario inside the method `testAddDDMStructure()` using `AssertUtils.assertFailure`. You can see an example on class `ObjectDefinitionLocalServiceTest.java`

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java [hunk 3]
```diff
@@ -286,6 +311,73 @@ public void testCopyStructure() throws Exception {
...
+			ddmStructure.getExternalReferenceCode(), ddmStructure.getGroupId());
+
+		Assert.assertNull(
+			_ddmStructureLocalService.fetchDDMStructureByExternalReferenceCode(
+				ddmStructure.getExternalReferenceCode(),
+				ddmStructure.getGroupId()));
+	}
+
+	@Test(expected = NoSuchStructureException.class)
+	public void testDeleteNonexistingDDMStructureByExternalReferenceCode()
```
Comment (guilhermedcamacho) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java:415]: You should cover this scenario inside the method `testDeleteDDMStructureByExternalReferenceCode()` using `AssertUtils.assertFailure`.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java [hunk 4]
```diff
@@ -892,6 +1002,33 @@ public void testSearchGlobalSiteStructure() throws Exception {
...
+		structure.setExternalReferenceCode(externalReferenceCode);
+
+		structure = updateStructure(structure);
+
+		Assert.assertEquals(
+			externalReferenceCode, structure.getExternalReferenceCode());
+	}
+
+	@Test(expected = DuplicateDDMStructureExternalReferenceCodeException.class)
+	public void testUpdateDDMStructureWithExistingExternalReferenceCode()
```
Comment (guilhermedcamacho) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-test/src/testIntegration/java/com/liferay/dynamic/data/mapping/service/test/DDMStructureLocalServiceTest.java:1061]: You should cover this scenario inside the method `testUpdateDDMStructure()` using `AssertUtils.assertFailure`.

**Review by guilhermedcamacho — CHANGES_REQUESTED** [review]
Please also change this [commit message](https://github.com/liferay-bpm/liferay-portal/pull/3784/commits/7d6c75eb914dedc3cb9151c7fb004b2f8c1a3919) to buildService.

**Comment by rijema** [comment]
@guilhermedcamacho please see https://github.com/liferay-bpm/liferay-portal/pull/3794

**Pull request closed without merging**